File: win_app_2/src/services/translation/translationCache.ts
/**
 * LRU cache for translation results. Avoids re-translating identical chunks
 * (common phrases: "yes", "ok", repeated questions).
 * Key = sha256(SOURCE→TARGET‖normalized text).
 */
export class TranslationCache {
  private readonly capacity: number
  private readonly storage = new Map<string, string>()

  constructor(capacity = 1000) {
    this.capacity = capacity
  }

  lookup(text: string, sourceLang: string | null | undefined, targetLang: string): string | null {
    const key = TranslationCache.makeKey(text, sourceLang, targetLang)
    const value = this.storage.get(key)
    if (value === undefined) return null
    // MRU: re-insert
    this.storage.delete(key)
    this.storage.set(key, value)
    return value
  }

  store(text: string, sourceLang: string | null | undefined, targetLang: string, translation: string): void {
    const key = TranslationCache.makeKey(text, sourceLang, targetLang)
    if (this.storage.has(key)) {
      this.storage.delete(key)
    } else if (this.storage.size >= this.capacity) {
      const oldest = this.storage.keys().next().value
      if (oldest !== undefined) this.storage.delete(oldest)
    }
    this.storage.set(key, translation)
  }

  clear(): void {
    this.storage.clear()
  }

  get size(): number {
    return this.storage.size
  }

  static makeKey(text: string, sourceLang: string | null | undefined, targetLang: string): string {
    const normalized = text.trim().toLowerCase()
    const source = (sourceLang ?? 'AUTO').toUpperCase()
    const target = targetLang.toUpperCase()
    return TranslationCache.sha256Sync(`${source}→${target}‖${normalized}`)
  }

  /**
   * Lightweight synchronous hash. We don't need cryptographic strength — just
   * a stable, collision-resistant key. FNV-1a over the UTF-8 bytes, doubled to
   * approximate 128-bit width.
   */
  private static sha256Sync(input: string): string {
    let h1 = 0x811c9dc5
    let h2 = 0x01000193
    for (let i = 0; i < input.length; i++) {
      const c = input.charCodeAt(i)
      h1 ^= c
      h1 = Math.imul(h1, 0x01000193)
      h2 ^= c + 0x9e3779b1
      h2 = Math.imul(h2, 0x85ebca6b)
    }
    const hex1 = (h1 >>> 0).toString(16).padStart(8, '0')
    const hex2 = (h2 >>> 0).toString(16).padStart(8, '0')
    return hex1 + hex2
  }
}

File: ios_app/QueenMamaIOS/Services/Translation/TranslationCache.swift
import Foundation
import CryptoKit

/// LRU cache for translation results to avoid re-translating identical chunks
/// (common in calls: "yes", "ok", "thank you", repeated questions).
/// Keys are sha256 over normalized source text + lang pair.
final class TranslationCache: @unchecked Sendable {
    private let capacity: Int
    private let lock = NSLock()
    private var storage: [String: String] = [:]
    private var order: [String] = []

    init(capacity: Int = 1000) {
        self.capacity = capacity
    }

    func lookup(text: String, sourceLang: String?, targetLang: String) -> String? {
        let key = Self.makeKey(text: text, sourceLang: sourceLang, targetLang: targetLang)
        lock.lock()
        defer { lock.unlock() }
        guard let value = storage[key] else { return nil }
        // Move key to end (most recently used)
        if let idx = order.firstIndex(of: key) {
            order.remove(at: idx)
            order.append(key)
        }
        return value
    }

    func store(text: String, sourceLang: String?, targetLang: String, translation: String) {
        let key = Self.makeKey(text: text, sourceLang: sourceLang, targetLang: targetLang)
        lock.lock()
        defer { lock.unlock() }
        if storage[key] != nil {
            // Already present — promote to MRU
            if let idx = order.firstIndex(of: key) {
                order.remove(at: idx)
            }
        } else if storage.count >= capacity {
            // Evict LRU
            if let oldest = order.first {
                order.removeFirst()
                storage.removeValue(forKey: oldest)
            }
        }
        storage[key] = translation
        order.append(key)
    }

    func clear() {
        lock.lock()
        defer { lock.unlock() }
        storage.removeAll()
        order.removeAll()
    }

    var count: Int {
        lock.lock()
        defer { lock.unlock() }
        return storage.count
    }

    static func makeKey(text: String, sourceLang: String?, targetLang: String) -> String {
        let normalized = text.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        let source = sourceLang?.uppercased() ?? "AUTO"
        let target = targetLang.uppercased()
        let composite = "\(source)→\(target)‖\(normalized)"
        let digest = SHA256.hash(data: Data(composite.utf8))
        return digest.compactMap { String(format: "%02x", $0) }.joined()
    }
}
